packages/contracts-bedrock/src/safe/DeputyPauseModule.sol

```diff
@@ -0,0 +1,207 @@
+    {
+        _setDeputy(_deputy, _deputySignature);
+        deputyGuardianModule = _deputyGuardianModule;
+        FOUNDATION_SAFE = _foundationSafe;
+        SUPERCHAIN_CONFIG = _superchainConfig;
+    }
+
+    /// @notice Getter function for the Foundation Safe address.
+    /// @return foundationSafe_ Foundation Safe address.
+    function foundationSafe() public view returns (Safe foundationSafe_) {
+        foundationSafe_ = FOUNDATION_SAFE;
+    }
+
+    /// @notice Getter function for the SuperchainConfig address.
+    /// @return superchainConfig_ SuperchainConfig address.
+    function superchainConfig() public view returns (ISuperchainConfig superchainConfig_) {
+        superchainConfig_ = SUPERCHAIN_CONFIG;
+    }
+
+    /// @notice Getter function for the Pause message typehash.
+    /// @return pauseMessageTypehash_ Pause message typehash.
+    function pauseMessageTypehash() public pure returns (bytes32 pauseMessageTypehash_) {
+        pauseMessageTypehash_ = PAUSE_MESSAGE_TYPEHASH;
+    }
+
+    /// @notice Getter function for the DeputyAuth message typehash.
+    /// @return deputyAuthMessageTypehash_ DeputyAuth message typehash.
+    function deputyAuthMessageTypehash() public pure returns (bytes32 deputyAuthMessageTypehash_) {
+        deputyAuthMessageTypehash_ = DEPUTY_AUTH_MESSAGE_TYPEHASH;
+    }
+
+    /// @notice Sets the deputy address.
+    /// @param _deputy Deputy address.
+    /// @param _deputySignature Deputy signature.
+    function setDeputy(address _deputy, bytes memory _deputySignature) external {
+        // Can only be called by the Foundation Safe itself.
+        if (msg.sender != address(FOUNDATION_SAFE)) {
+            revert DeputyPauseModule_NotFromSafe();
+        }
+
+        // Set the deputy address.
+        _setDeputy(_deputy, _deputySignature);
+    }
+
+    /// @notice Sets the DeputyGuardianModule.
+    /// @param _deputyGuardianModule DeputyGuardianModule address.
+    function setDeputyGuardianModule(IDeputyGuardianModule _deputyGuardianModule) external {
+        if (msg.sender != address(FOUNDATION_SAFE)) {
+            revert DeputyPauseModule_NotFromSafe();
+        }
+        deputyGuardianModule = _deputyGuardianModule;
+        emit DeputyGuardianModuleSet(_deputyGuardianModule);
+    }
+
+    /// @notice Calls the Foundation Safe's `execTransactionFromModuleReturnData()` function with
+    ///         the arguments necessary to call `pause()` on the Security Council Safe, which will
+    ///         then cause the Security Council Safe to trigger SuperchainConfig pause.
+    ///         Front-running this function is completely safe, it'll pause either way.
+    /// @param _nonce Signature nonce.
+    /// @param _signature ECDSA signature.
+    function pause(bytes32 _nonce, bytes memory _signature) external {
+        // Make sure the nonce hasn't been used yet.
+        if (usedNonces[_nonce]) {
+            revert DeputyPauseModule_NonceAlreadyUsed();
+        }
```

Comment:
Operationally, how you are thinking of nonces? Is the idea we just generate a random 32 byte nonce on the fly before signing and sending a pause, since in all likelihood that nonce will have never been used? Compared to the ERC-2612 approach where this is an incrementing nonce, and we'd need an extra call to fetch that nonce before signing and sending a pause

Reply:
Yep, basically. Just randomly generate 32 bytes. Would be astronomically unlikely that it's been used before.

Reply:
If we want, we can key the nonces by address so that different addresses don't need to worry about nonce reuse

Reply:
Keying by address feels like overkill honestly since collision odds are ~0 with random nonce generation each time, I'd prefer keeping it as-is for simplicity